xapian-letor/tests/api_letor.cc

```diff
@@ -861,3 +980,1088 @@ DEFINE_TESTCASE(tfdoclenfeature, generated)
+    TEST_EQUAL_DOUBLE(fvals_doc2[2], 0.0);
+
+    // test for weight in normalized form
+    TEST_EQUAL_DOUBLE(fvals_doc1[3], test_vals_doc1[3] / max_weight);
+    TEST_EQUAL_DOUBLE(fvals_doc2[3], test_vals_doc2[3] / max_weight);
+
+    return true;
+}
+
+DEFINE_TESTCASE(checkcreatefeaturevector_tfdoclencolllfcolllen, generated)
+{
+    vector<Xapian::Feature*> f;
+    Xapian::TfDoclenCollTfCollLenFeature *f1 =
+				new Xapian::TfDoclenCollTfCollLenFeature();
+    f.push_back(f1);
+
+    Xapian::FeatureList fl(f);
+    Xapian::Database db = get_database("apitest_createfeaturevector3",
+				       db_index_three_documents_no_common);
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("score"));
+    Xapian::MSet mset;
+    mset = enquire.get_mset(0, 10);
+
+    TEST(!mset.empty());
+
+    Xapian::QueryParser queryparser;
+    queryparser.set_stemmer(Xapian::Stem("en"));
+    queryparser.set_stemming_strategy(queryparser.STEM_ALL_Z);
+    queryparser.add_prefix("title", "S");
+    queryparser.add_prefix("description", "XD");
+
+    string querystring = "title:score description:score score";
+    Xapian::Query query = queryparser.parse_query(querystring);
+
+    auto fv = fl.create_feature_vectors(mset, query, db);
+    TEST_EQUAL(fv.size(), 2);
+    TEST_EQUAL(fv[0].get_fcount(), 4);
+    TEST_EQUAL(fv[1].get_fcount(), 4);
+
+    vector<double> fvals_doc1 = fv[0].get_fvals();
+    vector<double> fvals_doc2 = fv[1].get_fvals();
+    TEST_EQUAL(fvals_doc1.size(), 4);
+    TEST_EQUAL(fvals_doc2.size(), 4);
+
+    vector<double> test_vals_doc1(4);
+    test_vals_doc1[0] = 0.11394335230683678;
+    test_vals_doc1[1] = 0.76130720333102619;
+    test_vals_doc1[2] = 0.90738326700002048;
+
+    Xapian::MSetIterator it = mset.begin();
+    test_vals_doc1[3] = it.get_weight();
+
+    vector<double> test_vals_doc2(4);
+    // values will be same as that of the first document
+    test_vals_doc2[0] = 0.0;
+    test_vals_doc2[1] = 0.77758890362035493;
+    test_vals_doc2[2] = 0.78786362447009839;
+
+    ++it;
+    test_vals_doc2[3] = it.get_weight();
+
+    double max_val[4];
+    for (int i = 0; i < 4; ++i) {
+	max_val[i] = max(test_vals_doc1[i], test_vals_doc2[i]);
+    }
+
+    // test for title in normalized form
+    TEST_EQUAL_DOUBLE(fvals_doc1[0], test_vals_doc1[0] / max_val[0]);
+    TEST_EQUAL_DOUBLE(fvals_doc2[0], test_vals_doc2[0] / max_val[0]);
+
+    // test for body in normalized form
+    TEST_EQUAL_DOUBLE(fvals_doc1[1], test_vals_doc1[1] / max_val[1]);
+    TEST_EQUAL_DOUBLE(fvals_doc2[1], test_vals_doc2[1] / max_val[1]);
+
+    // test for whole in normalized form
+    TEST_EQUAL_DOUBLE(fvals_doc1[2], test_vals_doc1[2] / max_val[2]);
+    TEST_EQUAL_DOUBLE(fvals_doc2[2], test_vals_doc2[2] / max_val[2]);
+
+    // test for weight in normalized form
+    TEST_EQUAL_DOUBLE(fvals_doc1[3], test_vals_doc1[3] / max_val[3]);
+    TEST_EQUAL_DOUBLE(fvals_doc2[3], test_vals_doc2[3] / max_val[3]);
+
+    return true;
+}
```

Comment:
But presumably these new testcases are still useful?

Reply:
Oh, though many of these seem to be resubmitted as #252.

The ones that haven't seem to be colltfcolllenfeature (which was merged already in e94ccffe3ec5994ff20843dc9a84a1ed8890711e) and the populate*feature testcases which currently rely on the getters.  The latter would be useful to rework to use a user-created Feature subclass mechanism, but I think that would need implementing first.

Reply:
BTW, most English style guides seem to frown on using an apostrophe for the plural of an abbreviation (based on a web search for `abbreviation plural apostrophe`).  I'd suggest using `APIs` rather than `api's`.

Reply:
>  The latter would be useful to rework to use a user-created Feature subclass mechanism, but I think that would need implementing first.
>
Sure i will try that out.

Reply:
> But presumably these new testcases are still useful?

I think they are.

Reply:
> The latter would be useful to rework to use a user-created Feature subclass mechanism, but I think that would need implementing first.
> 
I don't see how we can do that without a public api connecting to the internal maps and I also don't understand the need to create a user-created Feature subclass.

Reply:
If we're aiming for a Feature subclassing API which is flexible enough to implement equivalents to the built-in features then it would be need at least a `protected` API.  But we'd be providing it for the purpose of allowing users to create their own features, rather than just exposing internal details purely for access by the test suite.

User-created features are useful for at least two categories of things:

* general purpose features which xapian-letor doesn't yet implement but would be candidates for merging into a future version.  It's useful (but not vital) to be able to implement these without having to patch and rebuild xapian-letor.  In some cases these would ideally need access to more stats so would need changes to xapian-letor anyway (e.g. if I wanted to add a feature which used the number of unique terms in the document).

* features which are specific to a particular application - for example if I'm implementing letor for search in an online store I might want to try adding features for price, number of sales, physical size of item, etc.

Many of the second category of features could actually be implemented without access to the stats (they'd probably either pull data from value slots in the Xapian DB or by querying some external data store).